src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1267,107 +1267,158 @@ public SolutionState WithDocumentFilePath(DocumentId documentId, string? filePat
         /// Creates a new solution instance with the document specified updated to have the text
         /// specified.
         /// </summary>
-        public SolutionState WithDocumentText(DocumentId documentId, SourceText text, PreservationMode mode = PreservationMode.PreserveValue)
+        public SolutionState WithDocumentText(
+            ImmutableArray<DocumentId> relatedDocuments,
+            SourceText text,
+            PreservationMode mode)
         {
-            var oldDocument = GetRequiredDocumentState(documentId);
-            if (oldDocument.TryGetText(out var oldText) && text == oldText)
+            var newState = this;
+            foreach (var documentId in relatedDocuments)
             {
-                return this;
+                var oldDocument = newState.GetRequiredDocumentState(documentId);
+                if (oldDocument.TryGetText(out var oldText) && text == oldText)
+                    continue;
+
+                newState = newState.UpdateDocumentState(oldDocument.UpdateText(text, mode), textChanged: true);
             }
 
-            return UpdateDocumentState(oldDocument.UpdateText(text, mode), textChanged: true);
+            return newState;
```

Comment:
this pattern repeats so much, it would probably make sense to have a helper.  What do you think @jasonmalinowski ?